### PR TITLE
feat[cartesian]: K offset write

### PR DESCRIPTION
**Pull request by FlorianDeconinck**
## Description

Looking at enabling offset write in K to help with physics.
Write is allowed on `FORWARD` and `BACKWARD`, disallowed for `PARALLEL`.

TODO:
- [x] Make tests for `conditional`
- [ ] Explore auto `extend` calculation
- [x] Fix `dace:X` backends

Link to #131 
Discussion happened on [GridTools concept ](https://github.com/GridTools/concepts/pull/34)

## Requirements

- [x] All fixes and/or new features come with corresponding tests.
- [x] Important design decisions have been documented in the approriate ADR inside the [docs/development/ADRs/](docs/development/ADRs/Index.md) folder.

### CODE REVIEW

File: src/gt4py/cartesian/frontend/gtscript_frontend.py [hunk 1]
```diff
@@ -1417,9 +1417,14 @@ def visit_Assign(self, node: ast.Assign) -> list:
         for t in node.targets[0].elts if isinstance(node.targets[0], ast.Tuple) else node.targets:
             name, spatial_offset, data_index = self._parse_assign_target(t)
             if spatial_offset:
```
Comment (havogt) [inline src/gt4py/cartesian/frontend/gtscript_frontend.py:1419]: are these the only changes to make it work? that's nice. It means we "only" have to sort out the conceptual problems

Comment (FlorianDeconinck) [reply src/gt4py/cartesian/frontend/gtscript_frontend.py:1419]: I was expecting more too! So far though, it's all it takes. Next I'll put in some integration tests with conditionals and I'll simplify Oliver's dual for-loop example he pulled from SHiELD microphysics to bring something we've encountered in the wild.

**Comment by FlorianDeconinck** [comment]
I've added a test that mimics the usage seen in the wild (column physics)

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py [hunk 2]
```diff
@@ -586,3 +586,79 @@ def test(out: Field[np.float64], inp: Field[np.float64]):
         backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 2), dtype=np.float64
     )
     test(out, inp)
+
+
+@pytest.mark.parametrize("backend", ["numpy"])
```
Comment (havogt) [inline tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:591]: Why only numpy?

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py [hunk 3]
```diff
@@ -586,3 +586,79 @@ def test(out: Field[np.float64], inp: Field[np.float64]):
...
+            B[0, 0, 1] = A
+
+    A = gt_storage.zeros(
+        backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 4), dtype=np.float64
+    )
+    A[:, :, :] = 42.0
+    B = gt_storage.zeros(
+        backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 4), dtype=np.float64
+    )
+    simple(A, B)
```
Comment (havogt) [inline tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:605]: I am a bit confused. Why is this not an oob write?

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py [hunk 4]
```diff
@@ -586,3 +586,79 @@ def test(out: Field[np.float64], inp: Field[np.float64]):
...
+        backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 4), dtype=np.float64
+    )
+    A[:, :, :] = 42.0
+    B = gt_storage.empty(
+        backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 4), dtype=np.float64
+    )
+    forward(A, B, 2.0)
+    # B will have non-updated values of A
+    assert (B[:, :, 1:3] == 42.0).all()
+
```
Comment (havogt) [inline tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:625]: let's add a check for A

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py [hunk 5]
```diff
@@ -586,3 +586,79 @@ def test(out: Field[np.float64], inp: Field[np.float64]):
...
+def test_K_offset_write(backend):
+    @gtscript.stencil(backend=backend)
+    def simple(A: Field[np.float64], B: Field[np.float64]):
+        with computation(FORWARD), interval(...):
+            B[0, 0, 1] = A
+
+    A = gt_storage.zeros(
+        backend=backend, aligned_index=(0, 0, 0), shape=(2, 2, 4), dtype=np.float64
+    )
+    A[:, :, :] = 42.0
```
Comment (havogt) [inline tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:601]: I didn't look what the quality of the other tests are, but maybe we can fill different values in k, just to make sure we actually see the shifted value, not the original one

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py (hunk 3 above)
Comment (FlorianDeconinck) [reply tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:605]: Extent calculation is doing the right job by excluding the first element from the loop

File: tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py (hunk 2 above)
Comment (FlorianDeconinck) [reply tests/cartesian_tests/integration_tests/multi_feature_tests/test_code_generation.py:591]: My bad, that was a debug setting, obviously it's all backends

**Comment by FlorianDeconinck** [comment]
Back to Draft PR:
- `dace:X` backends are not doing the right things, all other backends ✔️ 
- somehow `interval` gets properly calculated on a `(...)`? I was expecting failure. I'll dig more and try to convert findings in expected fails.

**Comment by FlorianDeconinck** [comment]
`dace:X` now working but that lead me to discover a bug with `While` and `dace:X` related to how we transform `FieldAccess` in the DaceIR

**Comment by FlorianDeconinck** [comment]
`while` loop bug unearthed in `dace:X` backends logged at: https://github.com/GridTools/gt4py/issues/1496

**Comment by FlorianDeconinck** [comment]
Poke @havogt, @edopao : I think this is ready for a round of review.

As said before it excludes `cuda` backend from the test (will be PR a deprecation on `cuda` backend). It also side step `dace:X` backends for one test due to a new bug discovered and logged.

**Review by havogt — COMMENTED** [review]
Note to myself: think about k-caching

**Comment by FlorianDeconinck** [comment]
`dace:X` backends repaired.

File: tests/cartesian_tests/definitions.py [hunk 6]
```diff
@@ -60,3 +61,13 @@ def _get_backends_with_storage_info(storage_info_kind: str):
 @pytest.fixture()
 def id_version():
     return gt_utils.shashed_id(str(datetime.datetime.now()))
+
+
+def _get_array_library(backend: str):
```
Comment (havogt) [inline tests/cartesian_tests/definitions.py:66]:
```suggestion
def get_array_library(backend: str):
```
let's make it public as it's used across modules

**Review by havogt — CHANGES_REQUESTED** [review]
Only minor things. I checked a few cases with k caches and they seem to work as expected. Can you review https://github.com/GridTools/concepts/pull/34?

**Comment by havogt** [comment]
cscs-ci run

**Review by havogt — APPROVED** [review]
thanks! looks good.

**Comment by havogt** [comment]
Summary of what I did so far:
- Reproduced the issue on my laptop with CUDA 11.2.2, can't reproduce with CUDA 12.5.
- Can't reproduce the issue in this example https://github.com/havogt/gridtools/blob/debug_offcenter_writer/tests/regression/offcenter_write.cpp even in the CI container on the same system (daint).

**Comment by FlorianDeconinck** [comment]
Back at this:
- `dace:gpu` generates plain wrong code (adding pointers?!)
- `gt:gpu` fails on CUDA 11 but 12

**Comment by FlorianDeconinck** [comment]
Can't reproduce the `dace:gpu` with CUDA 12

**Comment by FlorianDeconinck** [comment]
@havogt Can you cscs-run again, the ci died on a TIME LIMIT

**Comment by havogt** [comment]
cscs-ci run

**Pull request merged**